File: Bukkit/src/main/java/me/mrkirby153/kcutils/jukebox/NoteBlockSong.java
package me.mrkirby153.kcutils.jukebox;

import com.google.common.io.ByteArrayDataInput;
import com.google.common.io.ByteStreams;
import org.bukkit.Bukkit;
import org.bukkit.Sound;
import org.bukkit.entity.Player;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Objects;

public class NoteBlockSong {
    public static boolean debug = false;
    private final File location;
    boolean isDestroyed = false;
    private HashMap<Integer, ArrayList<NoteBlock>> notes = new HashMap<>();
    private double tempo;
    private int currentTick = 0;
    private Thread playerThread;
    private boolean isPlaying = true;
    private ArrayList<String> playersListening = new ArrayList<>();
    private String songName;

    public NoteBlockSong(File songLocation) {
        this.location = songLocation;
    }

    /**
     * "Tunes" the player to the song. Following notes will be played to the player
     *
     * @param player The player's name
     */
    public void addPlayer(String player) {
        if (playersListening.contains(player))
            return;
        playersListening.add(player);
        Player p = Bukkit.getPlayerExact(player);
        if (p == null)
            return;
        Bukkit.getPluginManager().callEvent(new SongStartedEvent(this, p));
    }

    /**
     * "Tunes" the player to the song. Following notes will be played to the player
     *
     * @param player The player object
     */
    public void addPlayer(Player player) {
        addPlayer(player.getName());
    }

    /**
     * Gets the song name (Loaded from the file)
     *
     * @return The song name
     */
    public String getSongName() {
        return this.songName;
    }

    /**
     * Temporarily pauses the playback of the song
     */
    public void pause() {
        isPlaying = false;
    }

    /**
     * Starts playback of the song
     */
    public void play() {
        loadSong();
        createThread();
    }

    /**
     * Untunes a player to the song
     *
     * @param player The player
     */
    public void removePlayer(Player player) {
        playersListening.remove(player.getName());
    }

    /**
     * Resumes playback of the song
     */
    public void resume() {
        isPlaying = true;
    }

    /**
     * Stops the playback of the song and destroys it
     */
    public void stop(boolean broadcastEvent) {
        isPlaying = false;
        currentTick = 0;
        if (broadcastEvent)
            Bukkit.getPluginManager().callEvent(new SongEndedEvent(NoteBlockSong.this));
        destroy();
    }

    /**
     * Creates the thread for playback
     */
    private void createThread() {
        this.playerThread = new Thread(() -> {
            while (!isDestroyed) {
                long startTime = System.currentTimeMillis();
                synchronized (NoteBlockSong.this) {
                    if (isPlaying) {
                        if (currentTick > notes.size()) {
                            isPlaying = false;
                            currentTick = 0;
                            Bukkit.getPluginManager().callEvent(new SongEndedEvent(NoteBlockSong.this));
                            destroy();
                            return;
                        }
                        int tick = currentTick++;
                        for (String s : playersListening) {
                            Player p = Bukkit.getPlayerExact(s);
                            playTick(p, tick);
                        }
                    }
                    long duration = System.currentTimeMillis() - startTime;
                    int delayMillis = getDelay() * 50;
                    if (duration < delayMillis) {
                        try {
                            Thread.sleep(delayMillis - duration);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                }
            }
        });
        this.playerThread.setName("JukeboxThread");
        this.playerThread.setPriority(10);
        this.playerThread.start();
        playersListening.stream().map(Bukkit::getPlayerExact).filter(Objects::nonNull).forEach(p -> Bukkit.getPluginManager().callEvent(new SongStartedEvent(this, p)));
    }

    /**
     * Destroy the playback thread
     */
    private void destroy() {
        isDestroyed = true;
    }

    /**
     * Gets the delay for scheduling song playback
     *
     * @return The delay
     */
    private short getDelay() {
        return ((short) (20 / tempo));
    }

    /**
     * Parses the NoteBlockStudioParser file and loads them into {@link NoteBlockSong#notes} list
     */
    private void loadSong() {
        try {
            byte[] bytes = Files.readAllBytes(location.toPath());
            ByteArrayDataInput input = ByteStreams.newDataInput(bytes);
            this.songName = input.readUTF();
            int totalChords = input.readInt();
            this.tempo = input.readDouble();
            for (int i = 0; i < totalChords; i++) {
                int tick = input.readInt();
                int notesInChord = input.readInt();
                ArrayList<NoteBlock> chord = new ArrayList<>();
                for (int j = 0; j < notesInChord; j++) {
                    byte instrument = input.readByte();
                    int pitch = input.readInt();
                    NoteBlock nb = new NoteBlock(tick, instrument, pitch);
                    chord.add(nb);
                }
                notes.put(tick, chord);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * Plays the corresponding tick to the player
     *
     * @param player The player to play the chord for
     * @param tick   The tick in the song to play
     */
    private void playTick(Player player, int tick) {
        ArrayList<NoteBlock> chord = this.notes.get(tick);
        if (debug)
            System.out.println("TICK: " + tick);
        if (chord != null) {
            for (NoteBlock nb : chord) {
                Sound sound = null;
                switch (nb.getInstrument()) {
                    case PIANO:
                        sound = Sound.BLOCK_NOTE_HARP;
                        break;
                    case DOUBLE_BASS:
                        sound = Sound.BLOCK_NOTE_BASS;
                        break;
                    case BASS_DRUM:
                        sound = Sound.BLOCK_NOTE_BASEDRUM;
                        break;
                    case SNARE_DRUM:
                        sound = Sound.BLOCK_NOTE_SNARE;
                        break;
                    case CLICK:
                        sound = Sound.BLOCK_NOTE_HAT;
                        break;
                }
                if (player == null)
                    return;
                if (debug)
                    System.out.println("\tNOTE: " + nb.toString());
                player.playSound(player.getLocation(), sound, 0.75f, (float) nb.getPitch().getNoteblockPitch());
            }
        }
    }
}

File: Bukkit/src/main/kotlin/me/mrkirby153/kcutils/jukebox/NoteBlockSong.kt
package me.mrkirby153.kcutils.jukebox

import com.google.common.io.ByteStreams
import org.bukkit.Bukkit
import org.bukkit.Sound
import org.bukkit.entity.Player
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.util.*

class NoteBlockSong(private val location: File) {
    private var isDestroyed = false
    private val notes = HashMap<Int, ArrayList<NoteBlock>>()
    private var tempo: Double = 0.toDouble()
    private var currentTick = 0
    private var playerThread: Thread? = null
    private var isPlaying = true
    private val playersListening = ArrayList<String>()
    /**
     * Gets the song name (Loaded from the file)
     *
     * @return The song name
     */
    var songName: String? = null
        private set

    /**
     * "Tunes" the player to the song. Following notes will be played to the player
     *
     * @param player The player's name
     */
    fun addPlayer(player: String) {
        if (playersListening.contains(player))
            return
        playersListening.add(player)
        val p = Bukkit.getPlayerExact(player) ?: return
        Bukkit.getPluginManager().callEvent(SongStartedEvent(this, p))
    }

    /**
     * "Tunes" the player to the song. Following notes will be played to the player
     *
     * @param player The player object
     */
    fun addPlayer(player: Player) {
        addPlayer(player.name)
    }

    /**
     * Temporarily pauses the playback of the song
     */
    fun pause() {
        isPlaying = false
    }

    /**
     * Starts playback of the song
     */
    fun play() {
        loadSong()
        createThread()
    }

    /**
     * Untunes a player to the song
     *
     * @param player The player
     */
    fun removePlayer(player: Player) {
        playersListening.remove(player.name)
    }

    /**
     * Resumes playback of the song
     */
    fun resume() {
        isPlaying = true
    }

    /**
     * Stops the playback of the song and destroys it
     */
    fun stop(broadcastEvent: Boolean) {
        isPlaying = false
        currentTick = 0
        if (broadcastEvent)
            Bukkit.getPluginManager().callEvent(SongEndedEvent(this@NoteBlockSong))
        destroy()
    }

    /**
     * Creates the thread for playback
     */
    private fun createThread() {
        this.playerThread = Thread {
            while (!isDestroyed) {
                val startTime = System.currentTimeMillis()
                synchronized(this@NoteBlockSong) {
                    if (isPlaying) {
                        if (currentTick > notes.size) {
                            isPlaying = false
                            currentTick = 0
                            Bukkit.getPluginManager().callEvent(SongEndedEvent(this@NoteBlockSong))
                            destroy()
                            return@synchronized
                        }
                        val tick = currentTick++
                        for (s in playersListening) {
                            val p = Bukkit.getPlayerExact(s)
                            playTick(p, tick)
                        }
                    }
                    val duration = System.currentTimeMillis() - startTime
                    val delayMillis = delay * 50
                    if (duration < delayMillis) {
                        try {
                            Thread.sleep(delayMillis - duration)
                        } catch (e: InterruptedException) {
                            e.printStackTrace()
                        }

                    }
                }
            }
        }
        this.playerThread!!.name = "JukeboxThread"
        this.playerThread!!.priority = 10
        this.playerThread!!.start()
        playersListening.map { Bukkit.getPlayerExact(it) }.filter { it != null }.forEach { Bukkit.getPluginManager().callEvent(SongStartedEvent(this, it)) }
    }

    /**
     * Destroy the playback thread
     */
    private fun destroy() {
        isDestroyed = true
    }

    /**
     * Gets the delay for scheduling song playback
     *
     * @return The delay
     */
    private val delay: Short
        get() = (20 / tempo).toShort()

    /**
     * Parses the NoteBlockStudioParser file and loads them into [NoteBlockSong.notes] list
     */
    private fun loadSong() {
        try {
            val bytes = Files.readAllBytes(location.toPath())
            val input = ByteStreams.newDataInput(bytes)
            this.songName = input.readUTF()
            val totalChords = input.readInt()
            this.tempo = input.readDouble()
            for (i in 0 until totalChords) {
                val tick = input.readInt()
                val notesInChord = input.readInt()
                val chord = ArrayList<NoteBlock>()
                for (j in 0 until notesInChord) {
                    val instrument = input.readByte()
                    val pitch = input.readInt()
                    val nb = NoteBlock(tick, instrument, pitch)
                    chord.add(nb)
                }
                notes.put(tick, chord)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    /**
     * Plays the corresponding tick to the player
     *
     * @param player The player to play the chord for
     * @param tick   The tick in the song to play
     */
    private fun playTick(player: Player?, tick: Int) {
        val chord = this.notes[tick]
        if (debug)
            println("TICK: " + tick)
        if (chord != null) {
            for (nb in chord) {
                var sound: Sound? = null
                when (nb.instrument) {
                    NoteBlock.Instrument.PIANO -> sound = Sound.BLOCK_NOTE_HARP
                    NoteBlock.Instrument.DOUBLE_BASS -> sound = Sound.BLOCK_NOTE_BASS
                    NoteBlock.Instrument.BASS_DRUM -> sound = Sound.BLOCK_NOTE_BASEDRUM
                    NoteBlock.Instrument.SNARE_DRUM -> sound = Sound.BLOCK_NOTE_SNARE
                    NoteBlock.Instrument.CLICK -> sound = Sound.BLOCK_NOTE_HAT
                }
                if (player == null)
                    return
                if (debug)
                    println("\tNOTE: " + nb.toString())
                player.playSound(player.location, sound, 0.75f, nb.pitch.noteblockPitch)
            }
        }
    }

    companion object {
        var debug = false
    }
}
